### PR TITLE
ramips: fix RBMxxG name and partitionning

### PR DESCRIPTION
**Pull request by f00b4r0**
This patch series sets the correct partition scheme for the RouterBoot section of the flash on these devices.

This section is subdivided in several segments, as they are on ar71xx
RB devices, albeit with different offsets and sizes.

The device name is corrected to match the hardware-stored (in hard_config)
device name.

**the RBM33G is already included in 18.06, the corresponding patch should be cherry-picked.**

### CODE REVIEW

File: target/linux/ramips/dts/RBM11G.dts [hunk 1]
```diff
@@ -90,17 +90,42 @@
...
+		// XXX empiric value to obtain actual 10MHz SCK at the chip
 		spi-max-frequency = <3125000>;
 
 		partition@0 {
+			label = "RouterBoot";
+			reg = <0x0 0x40000>;
+			read-only;
+		};
+
+		partition@1 {
```
Comment (mkresin) [inline target/linux/ramips/dts/RBM11G.dts]: has to be partition@0 according to the reg.

**Review by mkresin — CHANGES_REQUESTED** [review]

**Comment by f00b4r0** [comment]
Updated set produce no DTB error for the partition scheme.
@TobleMiner could you please test the latest patch set and confirm the mtd partitioning is correct?
Thanks.

**Comment by TobleMiner** [comment]
@f00b4r0 Hey, took me some time but I just tested your changes on both the RBM33G and the RBM11G. New partition layout looks ok. I've attached it for both the RBM33G and the RBM11G sou you can double check it's what you'd expect:

RBM33G:
```
[    2.343572] m25p80 spi0.0: w25x40 (512 Kbytes)
[    2.348055] 6 fixed-partitions partitions found on MTD device spi0.0
[    2.354435] Creating 6 MTD partitions on "spi0.0":
[    2.359213] 0x000000000000-0x000000040000 : "RouterBoot"
[    2.365427] 0x000000000000-0x00000000f000 : "routerboot"
[    2.371471] 0x00000000f000-0x000000010000 : "hard_config"
[    2.377692] 0x000000010000-0x00000001f000 : "routerboot2"
[    2.383903] 0x000000020000-0x000000021000 : "soft_config"
[    2.390072] 0x000000030000-0x000000031000 : "bios"
[    2.406544] m25p80 spi0.1: w25q128 (16384 Kbytes)
[    2.411275] 2 fixed-partitions partitions found on MTD device spi0.1
[    2.417626] Creating 2 MTD partitions on "spi0.1":
[    2.422421] 0x000000000000-0x000000040000 : "RouterBootFake"
[    2.428931] 0x000000040000-0x000001000000 : "firmware"
[    2.521945] 2 minor-fw partitions found on MTD device firmware
```

RBM11G:
```
[    2.466697] m25p80 spi0.0: w25q128 (16384 Kbytes)
[    2.471450] 7 fixed-partitions partitions found on MTD device spi0.0
[    2.477826] Creating 7 MTD partitions on "spi0.0":
[    2.482604] 0x000000000000-0x000000040000 : "RouterBoot"
[    2.488948] 0x000000000000-0x00000000f000 : "routerboot"
[    2.495289] 0x00000000f000-0x000000010000 : "hard_config"
[    2.501596] 0x000000010000-0x00000001f000 : "routerboot2"
[    2.507966] 0x000000020000-0x000000021000 : "soft_config"
[    2.514307] 0x000000030000-0x000000031000 : "bios"
[    2.520108] 0x000000040000-0x000001000000 : "firmware"
[    2.597143] 2 minor-fw partitions found on MTD device firmware
```

The mac adresses are loaded from flash correctly, too.

**Comment by f00b4r0** [comment]
@TobleMiner Thank you very much for testing, this confirms the changes work as expected!

**Comment by mkresin** [comment]
Following our discussion in IRC, I've to admit I'm impressed by your hacker approach. Workaround the outlined issue, by keeping the overlapping partition.

Nevertheless, it will not be committed this way. It took me way to long to get ramips in shape by removing all the "special cases" and I'm not going to introduce similar hacks which makes ar71xx that ugly looking.

Since I appreciate your work and this PR, I followed your advice to do the changes on my own. I've modified your commits and merged the result to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary).

Please ACK/NAK the changed commits.

**Comment by f00b4r0** [comment]
I will not support your changes. I will not provide further support for this code and I will not work on fixing rbcfg for this platform.

I take note that you arbitrarily make choices according to your own preferences instead of sound technical arguments: I fail to see how that follows the consensus rule of OpenWRT.

**Comment by f00b4r0** [comment]
@mkresin please remove my signed-off-by from these commits.

**Comment by f00b4r0** [comment]
Another argument in favor of the top level partition (besides making clear this whole area, holes included, are covered by OEM; and making it easy to request dumps and create a database of said dumps) is that since the top partition has the same name as the OEM code, it might make it possible to use OEM tools in LEDE.

**Comment by mkresin** [comment]
As we can't find a consensus on the topic, I'll drop the commits from my staging tree and close the PR.

**Comment by jow-** [comment]
@mkresin - can you please in layman's terms what the problem is with the proposed changes?

**Comment by f00b4r0** [comment]
I'll restate the key reasons for the proposed scheme:
- OEM defines the `0x0 0x40000` area as 'RouterBoot' in their source:
  - This whole segment should thus be considered "reserved by OEM"
  - Subsegments are reversed engineered and may change
  - Holes between segments may get filled eventually
  - OEM tools might be usable with this top level partition present and named identically with OEM
  - The top level partition is a convenient way to ask for a user dump ("please send the content of the RouterBoot partition") and build a database of such dumps
  - I expect it might eventually be possible to split that top level partition at runtime with a splitter
- We use top levels partitions everywhere ('firmware' is one, 'rootfs' is another)
- It doesn't break anything

And if that matters in any way: *I* am the one offering to support this code and make the necessary changes to e.g. rbcfg so that it can work with this platform.

**Comment by f00b4r0** [comment]
FWIW the way I defined these partitions in the DTS is the _canonical_ way to define fixed partitions as described in https://www.kernel.org/doc/Documentation/devicetree/bindings/mtd/partition.txt

That it's not used elsewhere is 1) inaccurate (all bcm targets use it, ath79 as well, ipq and even lantiq; and 2) for the DTSes that don't define partitions this way, I'd contend this might well be regarded as a bug.

**Comment by mkresin** [comment]
> Hm, I always thought that overlapping partitions are quite normal - e.g. firmware (kernel + rootfs) but I only have limited experience with ar71xx. What is undesired about them?

The firmware partition splitting thingy is done by the kernel you don't define kernel, rootfs + firmware in the device tree source files. Even for ar71xx where both is defined, it is usually wrong/not needed. The sysupgrade image is always written to the firmware partition and you need to ensure that your kernel is padded to match the size of the kernel partition. Hence, where it is spotted, it gets in most cases fixed.

> We use top levels partitions everywhere ('firmware' is one, 'rootfs' is another)

No we don't as outlined above. We usually don't have firmware along a rootfs + kernel partition. There might be exceptions, which doesn't mean it is right.

> It doesn't break anything

No, it introduces just another special case which could be a pain if you are doing mass operations. Been there, done that.

**Comment by mkresin** [comment]
> OEM defines the 0x0 0x40000 area as 'RouterBoot' in their source: 

What I might to understand, the 'RouterBoot' partition has an internal structure/partition map we are not able to parse and therefore we can't split the partition according to the internal partition map. Hence we trust that it never changes and add the sub parts as fixed partitions?

What I never asked is why do we need to have the subpartitions defined. It looks like it worked with one big 'RouterBoot' partition so far.

**Comment by f00b4r0** [comment]
> What I might to understand, the 'RouterBoot' partition has an internal structure/partition map we are not able to parse and therefore we can't split the partition according to the internal partition map. Hence we trust that it never changes and add the sub parts as fixed partitions?

We are able to parse it to some extent: that's how I figured the currently hardcoded boundaries. The structure is common across MikroTik hardware and Gabor Juhos wrote the basis of a very minimal (and incomplete) parser that's used successfully to extract the subpartitions on ar71xx (see example in my mach-rbspi.c code for instance).

And before you demand I write a proper splitter instead of hardcoding the partitions, I'll point out two things:

- We first need to collect several dumps from different hardware kits it to make it statistically possible to reverse-engineer a reliable splitter.
- I contribute pro-bono out of good will and free time. I don't want to commit to writing this splitter now and you don't get to require that from me as a prerequisite to accepting _working, correct and in-style_ code.

Finally, if the top level partition is present _now_ it will pave the way for the splitter. If that partition is absent (as it is on ar71xx, sadly), it closes the door to the splitter (if you aren't familiar with mtdsplit code and operation, it does need a top level partition to split from and enforce boundaries within. I'll remind you I had the opportunity to learn mtdsplit internals at length when I wrote mtdsplit-minor, another time with your "rigorous input" that's not a very pleasant memory for me).

And heaven forbid what would happen if someone found it a good idea to later "reclaim" the unassigned holes for use as user storage, another thing the top level partition protects again.

> What I never asked is why do we need to have the subpartitions defined.

Interesting shift: after objecting to the top level partition you now object to the sublevel ones. Are you trying to exhaust me by having a circular argument?

Anyway, that (valid) question probably stems as consequence of you not knowing the underlying hardware, how it's handled in OpenWRT, and trying instead to impose your views/preferences on someone who went through the pain of learning said hardware to the best of his abilities and poring through the 12+ megabytes of obfuscated OEM code; all that while trying very hard to follow OpenWRT's best/usual practices and to offer quality (by your own admission) contributions.

> It looks like it worked with one big 'RouterBoot' partition so far.

It didn't: that partition didn't exist and there were two partitions (firmware excluded) that did not correspond to any actual layout on the flash.

The subpartitions are necessary to use e.g. rbcfg on OpenWRT. Currently rbcfg cannot work on ramips because the 'soft_config' subpartition is missing (and because it has hardcoded endianness, but that's another story).

The 'hard_config' partition contains, among other things, the MAC addresses of the hardware and the model name string (which we use on ar71xx, not on ramips), the serial number, etc.

(I look forward to your suggestion that we need none of the OEM-related subpartitions defined in many DTSes and that we can simply hardcode the offsets of the data we need from 0x0 :P)

Joking aside, I think this is a good summary of our differences: I'm trying to demonstrate that I know what I'm talking about by providing reasonable technical and factual elements, answering your objections the same way and keeping an open mind for technically sound rebuttals; when you choose to speak in absolutes backed solely by your own preferences without ever technically addressing my arguments. And unless I'm mistaken, you're not trying to enforce these preferences on the premises that you will take over supporting this code/hardware either, so you're essentially asking me to do things your way _and_ deal with it.

I'll rest my case for today: it's late :P

**Comment by blogic** [comment]
tl;dr ... post it on the ML if you want people to read it and come to a conclusion whether we are ok or not with having overlapping partitions.

**Comment by f00b4r0** [comment]
posted.
